Source/WebCore/page/WebKitStringMatcher.cpp

```diff
@@ -0,0 +1,272 @@
+        ASSERT(nodeToStateIndex.contains(&node));
+        size_t stateIndex = nodeToStateIndex.get(&node) + node.lengths.size();
+        ASSERT(states[stateIndex].transitionsBeginIndexOrMatchIdentifier == unlinkedSentinel);
+        ASSERT(states[stateIndex].transitionsEndIndexOrMatchSentinel == unlinkedSentinel);
+        if (transitionsBegin > std::numeric_limits<uint16_t>::max() || transitionsEnd > std::numeric_limits<uint16_t>::max())
+            return false;
+        states[stateIndex] = WebKitStringMatcher::State {
+            static_cast<uint16_t>(transitionsBegin),
+            static_cast<uint16_t>(transitionsEnd)
+        };
+        return true;
+    });
+    if (!success)
+        return std::nullopt;
+
+#if ASSERT_ENABLED
+    for (auto& state : states) {
+        ASSERT(state.transitionsBeginIndexOrMatchIdentifier != unlinkedSentinel);
+        ASSERT(state.transitionsEndIndexOrMatchSentinel != unlinkedSentinel);
+    }
+#endif
+
+    return { { WTFMove(states), WTFMove(transitions) } };
+}
+
+std::optional<Vector<uint16_t>> WebKitStringMatcher::dfaForMatchingStrings(const Vector<std::pair<String, uint16_t>>& vector)
+{
+    Trie trie(vector);
+    auto dfa = trie.serialize();
+    if (!dfa)
+        return std::nullopt;
+
+    size_t statesSize = dfa->states.size();
+    size_t transitionsSize = dfa->transitions.size();
+    RELEASE_ASSERT(statesSize <= std::numeric_limits<uint16_t>::max());
+    RELEASE_ASSERT(transitionsSize <= std::numeric_limits<uint16_t>::max());
+
+    Vector<uint16_t> data;
+
+    constexpr size_t headerSize { 4 };
+    data.reserveInitialCapacity(headerSize + statesSize * 2 + transitionsSize * 2);
+
+    data.append(0);
+    data.append(0);
+    data.append(statesSize);
+    data.append(transitionsSize);
+
+    for (auto& state : dfa->states) {
+        data.append(state.transitionsBeginIndexOrMatchIdentifier);
+        data.append(state.transitionsEndIndexOrMatchSentinel);
+    }
+    for (auto& transition : dfa->transitions) {
+        data.append(transition.codeUnitToCheck);
+        data.append(transition.stateIndexToTransitionToIfMatched);
+    }
+
+    return { WTFMove(data) };
+}
+
+std::optional<std::pair<std::span<const WebKitStringMatcher::State>, std::span<const WebKitStringMatcher::Transition>>> WebKitStringMatcher::statesAndTransitionsFromVersionedData(std::span<const uint8_t> bytes)
+{
+    static_assert(sizeof(State) == sizeof(Transition));
+    static_assert(sizeof(State) == sizeof(uint32_t));
+    if (bytes.size() % sizeof(State))
+        return std::nullopt;
+    auto integers = spanReinterpretCast<const uint32_t>(bytes);
+    constexpr uint32_t currentVersion { 0 };
+    constexpr size_t integersInHeader { 2 };
+    if (integers.size() < integersInHeader
+        || integers[0] != currentVersion)
+        return std::nullopt;
+    uint32_t secondInteger = integers[1];
+    size_t stateLength = static_cast<uint16_t>(secondInteger);
```

Comment:
We should probably add a FIXME saying that this code needs some care if it were to be used in a big endian CPU.
Better yet, maybe spit out an error when compiling with big endian CPU. Alternatively, we can just add `bigEndianToLittleEndian2` call in `#if CPU(BIG_ENDIAN)`

Reply:
I'll just fail completely on big-endian CPUs.  If someone wants to implement that, they can.